sdk/python/lib/pulumi/resource.py

```diff
@@ -160,13 +160,28 @@ def __init__(self,
                         self._providers = {**self._providers, pkg: provider}
 
         if not custom:
-            providers = opts.providers
-            if providers is not None:
-                self._providers = {**self._providers, **providers}
+            providers = self.__convert_providers(opt.provider, opts.providers)
+            self._providers = {**self._providers, **providers}
 
         self._protect = bool(opts.protect)
         register_resource(self, t, name, custom, props, opts)
 
+    def __convert_providers__(self, provider: Optional['ProviderResource'], providers: Union[Mapping[str, 'ProviderResource'],List['ProviderResource']]) -> Mapping[str, 'ProviderResource']
```

Comment:
Double-underscore names are reserved for Python intrinsics - `_convert_providers` instead? (A single underscore indicates private-ness of a function, generally)

Reply:
yup yup